llvm/docs/NVPTXUsage.rst

```diff
@@ -311,7 +311,37 @@ The ``@llvm.nvvm.fence.proxy.tensormap_generic.*`` is a uni-directional fence us
   ``@llvm.nvvm.fence.proxy.tensormap_generic.acquire.*`` ``fence.proxy.tensormap::generic.acquire.* [addr], size``
   ====================================================== =========================================================
 
-The address operand ``addr`` and the operand ``size`` together specify the memory range ``[addr, addr+size)`` on which the ordering guarantees on the memory accesses across the proxies is to be provided. The only supported value for the ``size`` operand is ``128`` and must be an immediate. Generic Addressing is used unconditionally, and the address specified by the operand addr must fall within the ``.global`` state space. Otherwise, the behavior is undefined. For more information, see `PTX ISA <https://docs.nvidia.com/cuda/parallel-thread-execution/#parallel-synchronization-and-communication-instructions-membar>`_.
+The address operand ``addr`` and the operand ``size`` together specify the memory range ``[addr, addr+size)`` on which the ordering guarantees on the memory accesses across the proxies is to be provided. The only supported value for the ``size`` operand is ``128`` and must be an immediate. Generic Addressing is used unconditionally, and the address specified by the operand addr must fall within the ``.global`` state space. Otherwise, the behavior is undefined. For more information, see PTX ISA `<https://docs.nvidia.com/cuda/parallel-thread-execution/#parallel-synchronization-and-communication-instructions-membar>`_.
+
+'``llvm.nvvm.tensormap.cp_fenceproxy.global.shared.tensormap_generic.release.*.sync.aligned``'
```

Comment:
This deserves some sort of prize for the longest intrinsic name I've see so far. NVIDIA's PTX went a bit overboard with the instruction names lately, IMO. 

I'm usually a proponent of intrinsics closely matching the instruction name, but I'm not sure that it's the best approach here. These are just way too long for no good practical reason I can see. 

I wonder if we may come up with a more concise naming scheme for these intrinsics. 
If we expect to see only limited set of parameter combinations, perhaps we can just skip them. I.e. if `global.shared` is the only supported AS combination, we could just document it and remove it from the intrinsic name.

Reply:
> I wonder if we may come up with a more concise naming scheme for these intrinsics.

While it may be possible to design generic intrinsics that "move" some of the modifiers to function arguments, doing that later is compatible with 1:1 PTX exposure today, and it would require quite a bit more care (to avoid adding an intrinsic that would require a breaking change to support newer PTX versions).

Reply:
I don't mean one generic intrinsic handling everything, but rather to skip mentioning parts that never change.

E.g. until we have variants that support any other accesses other than `global.shared`, there's no benefit in encoding them into the name. 1:1 mapping is still in place.

If/when we will need to introduce other variants, e.g. `shared.shared`, then we can add intrinsics with the address space encoded in the name, and auto-upgrade the abbreviated one to the `global.shared`.

Another option is to use a less verbose mnemonics. E.g. instead of `llvm.nvvm.tensormap.cp_fenceproxy.global.shared.tensormap_generic.release.*.sync.aligned` we could use `llvm.nvvm.tm.cfp.g.s.tmg.release.*.sync.aligned`

Right now, according to the PTX spec, a lot fo the instruction name components are fixed:

```
tensormap.cp_fenceproxy.cp_qualifiers.fence_qualifiers.sync.aligned  [dst], [src], size;

.cp_qualifiers    = { .global.shared::cta }
.fence_qualifiers = { .to_proxy::from_proxy.release.scope }
.to_proxy::from_proxy  = { .tensormap::generic }
.scope            = { .cta, .cluster, .gpu , .sys }
```

Once this convoluted structure is expanded, only the `.scope` value appears to change in the name.
One can guess, based on this syntax structure, that NVIDIA may have plans to extend it in the future, but without nothing else to go on, it's hard to predict how, when, or whether it will actually change.

IMO, a naming scheme with abbreviated or omitted fields would be somewhat more user friendly.

That said, it's a cosmetic issue. While I'm not happy with the excessively long names, I can live with them.

Reply:
> If/when we will need to introduce other variants, e.g. shared.shared, then we can add intrinsics with the address space encoded in the name, and auto-upgrade the abbreviated one to the global.shared.

What do you mean by "auto-upgrade"? That new variants are added, but variants without modifiers are kept with their pre-existing meaning?

Reply:
LLVM has a way to convert old/deprecated intrinsics it sees on IR input to their current counterparts: llvm/lib/IR/AutoUpgrade.cpp

Reply:
If we are choosing between the options of abbreviating or omitting fields, I would vote for omitting. You suggested dropping `global.shared`. Should we also drop `sync.aligned` and just define it this way as this is always `sync` and `aligned`?

Reply:
See above:
> That said, it's a cosmetic issue. While I'm not happy with the excessively long names, I can live with them.

That said, I'll address your points
> For those that do not, parts of the longer names (or call arguments) highlight deviation from defaults (generic, release, .sync.aligned), and help with search ability in the PTX spec, which at the end is the reference documentation for these intrinsics. There may be future tooling advantages to consider (like LLVM->PTX verification, auto-generating some of these from a machine readable PTX spec, etc.).

A comment next to the intrinsic implementation, spelling out the exact instruction it will lower to addresses the searchability issue without cluttering the IR with excessively long names. The names are not intended to spell out all the details in full sentences, we've got documentation for that. The purpose of the intrinsic is to provide understandable consistent mapping for the underlying instruction. Whether we can come up with such mapping that's better than the default exact map is what I'm trying to figure out. Naming is hard.

> I feel like we could debate what to do here forever, so maybe we should zoom out a bit: this is not the longest instruction there is, and it is "short" compared with some PTX instructions that may be coming.

"It can be worse" is a weak counterargument. The "it *will* be worse" part does concern me a bit and suggests that we should try to come up with a better name, because those longer instructions will percolate into LLVM IR. While we can't do much about NVIDIA's choices of instruction names, we do not have to follow them, if we can do better.

> We could spend engineering effort trying to shorten every intrinsic for every new PTX instruction,

We're not doing that. As I said above, I'm all for straightforward 1:1 mapping, but this specific case (and as you mention incoming instructions) looks like we may need to consider other options before we proceed. We probably should discuss it on LLVM's discourse forum. If there's general consensus that N-line long names are fine, so be it.

> - slows down support for new HW in NVPTX (NVPTX is missing hundreds of intrinsics), and

Rushing is rarely a good choice, particularly when there are open questions, and no existing users for the new intrinsics. Specifically for intrinsics, delay with their implmentation is almost never a showstopper, as there's always an option to use inline asm as a workaround. So, yes, resolving the design issues/concerns is part of the review process, and it does take time, and no, the delay is not a good argument to short-circuit the review process.

> - risks spending more engineering effort once PTX adds a new modifier to unshorten them, then implementing AutoUpgrade, and maybe upgrading frontends.

What is your point? Regardless of what we do now, we will need to make changes in the future, refactor existing code, etc. Autopupgrade changes are largely mechanical and not a big deal in practice. Besides, they may or may not be needed, depending on which naming scheme we settle on, so this point is something to consider when we make that decision.

> I think it may actually be more valuable to, instead, prioritize adding support for new HW to NVPTX as soon as possible,

This is a false dichotomy, IMO. Yes, we need to add support for the new instructions and corresponding intrinsics, but we also need to consider how we do that. It's not an either/or proposition. We need both.

Reply:
Agree that we should move this to the forums, since this discussion goes beyond the scope of this one intrinsic and more into strategy.
Trying to get this back on track to discuss the spelling changes that we could do for this one intrinsic: 

- Shortening  intrinsic by removing `.global.shared`: I think this is fine as long as we raise an user error if the `addrspace` of the `src` and `dst` pointers do fullfill the requirements, to prevent undefined behavior.
- Shortening intrinsic by removing `.sync.aligned`: I think that incorrect use of `.sync.aligned` is one of the main sources of undefined behavior in LLVM IR. I don't know to which degree we can produce an user error if this requirement is not full-filled, but if we cannot reliably error I think there needs to be something in the intrinsic name that conveys that this is a `.sync.aligned` intrinsic. Does LLVM IR have a better name for this? (`convergent` seems to be subtly different)
- Shortening intrinsic by removing `.release`: this intrinsic performs multiple operations, one of which is a `fence` instruction. When a memory ordering is omitted, the default is not `release` (but `seq_cst` in C++, and `acq_rel` in PTX), which differ from the ordering used by this fence. The ordering reveals with which other memory instructions this one instruction builds a release pattern with, so I think we should keep it. If we wanted to shorten it, an option is to provide an `i32 x` argument that conveys the order (`i32 5`) is shorter than `.release`.
- Shortening intrinsic by removing `.tensormap_generic`. The fence in this instruction is a cross-proxy fence (`fence.proxy`) from `generic`-proxy to `tensormap`-proxy, and in contrast with other cross-proxy fence instructions (e.g. `fence.proxy.alias`) it is not a bidirectional `acq_rel` cross-proxy fence, but an uni-directional one from `generic`-proxy to `tensormap`-proxy only. `.tensormap_generic` brings all of this with it. My preference is that if we wanted to shorten it, we could add two `i32 dstp, i32 srcp` that specify the destination and source proxies for uni-directional proxy fences (for bi-directional proxy fences we could have just one `i32`).
- Shortening the scope: using `i32 NVPTX::Sco` instead.

So applying the shortenings, it could look like: 

```llvm
llvm.nvvm.tensormap.cp_fenceproxy.sync.aligned(
  ptr addrspace(global) %0, // dst
  ptr addrspace(shared) %1, // src
  i32 128, // size imm
  i32 5, // release
  i32 1, // tensormap-proxy
  i32 0, // generic-proxy
  i32 sco, // NVPTX::Scope
)
```

I think this exposure is still a 1:1 PTX exposure (and therefore per the prior comment I think the pros overweight the cons), it just mechanically picks a different syntax than the one this PR picks:
- Use `addrspace` instead of `.statespace` modifiers.
- Use `NVPTX::Order` enum instead of `.sem` modifier.
- Use  (a new) `NVPTX::Proxy` enum instead of mapping `.to_proxy::from_proxy` to `.to_proxy_from_proxy`. 
- Use `NVPTX::Scope` enum instead of a `.scope` modifier.